### DISCUSSION TITLE
Array initialization issues

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#1360] [Language Ideas]
We can initialize the array like this:
`int[] Arr = { 1, 2, 3, 4 };`
So, why we can't send an array parameter like this?
`var v = new Vector<int>({ 1, 2, 3, 4 });`
It gives the error:
> Cannot initialize type 'Vector<int>' with a collection initializer because it does not implement 'System.Collections.IEnumerable'

so we must do this:
`var v = new Vector<int>(new int[]{ 1, 2, 3, 4 });`

Why don't C# do the trick and convert 
`{ 1, 2, 3, 4 }`
to 
`new int[]{ 1, 2, 3, 4 }`
based on the type of the parameter?
Visual Basic Does this, and this code will compile:
`Dim v = New Vector(Of Integer)({1, 2, 3, 4})`

The same issue here:
`var Names = { "Ahmad", "Ali", "Abdo" };`
This will not compile, becuase C# can't infere that Names is a string array!
This is accepted in VB.NET:
`Dim Names = {"Ahmad", "Ali", "Abdo"}`

**Comment by bondsbw** [comment]
First to be clear, that syntax doesn't work even for `List<T>`:
```c#
var l = new List<int>({ 1, 2, 3, 4 });  // Error, initializer can't be enclosed in parentheses
```

It must be
```c#
var l = new List<int> { 1, 2, 3, 4 };
```

And why that isn't supported for `Vector<T>` is answered [here](https://stackoverflow.com/a/24033810/979199):

> You can't create immutable collection with a collection initializer because the compiler translates them into a sequence of calls to the `Add` method.

`Vector<T>` has no `Add` method.

**Comment by bondsbw** [comment]
With that out of the way, the proposal is related to #414 (Dictionary literals) and #1238 (JSON-like syntax for dictionary/array initialization and operation).

Also somewhat related to #100 ("Target-typed `new` expression").

**Comment by svick** [comment]
@bondsbw I think the array initializer is enclosed in parentheses on purpose. It's not supposed to be collection initializer, it's supposed to invoke the constructor that takes an array, just with a more succinct syntax than `new Vector<int>(new[] { 1, 2, 3, 4 })`.

**Comment by CyrusNajmabadi** [comment]
This needs an actual design.  For example, is ```var v = new Vector<int>({ 1, 2, 3, 4 });``` supposed to be shorthand for:

```var v = new Vector<int>(new [] { 1, 2, 3, 4 });```?  or does it mean:
```var v = new Vector<int>(new target_array_element_type [] { 1, 2, 3, 4 });```

?

With C#, when you write ```double[] d = { 1, 2, 3, 4 };``` you're getting an array of doubles.  i.e. it is not equivalent to ```double d= new [] { 1, 2, 3, 4 };``` which would fail (because you were assigning an array of ints to an array of doubles).

**Comment by CyrusNajmabadi** [comment]
I think all the examples so far that use untyped array creation (i.e. "new[]") are likely *not* what we'd want to do in the language.

**Comment by bondsbw** [comment]
@MohammadHamdyGhanem I'm confused, @svick makes a good point... you aren't talking about initializer syntax so where did this come from?

> Cannot initialize type 'Vector' with a collection initializer because it does not implement 'System.Collections.IEnumerable'

**Comment by CyrusNajmabadi** [comment]
@alrz why teh confusion?  It seems like it would just not be what you wanted.  If you had:

```c#
void Foo(double[] d) { ... }

Foo({ 1, 2, 3, 4 });
```

it seems like you'd want htat to be a new double array, not a new int array.  If we treated that expression as "new [] { ... }" then it would be a new int array.

**Comment by alrz** [comment]
@CyrusNajmabadi Ah, now I see what you meant. it should be definitely target-typed (just like local inits). IMO making it like `new[]` will cause it to fail most of the time due to some conversion error.. I'd like this to also work with `List<T>`, the question is that in the presence of `new()` and `new[]` would it worth it?

**Comment by ghost** [comment]
@bondsbw 
`var v = new Vector < int> ({ 1, 2, 3, 4 });`
Causes 4 errors, 3 of them about missing (, ; and )
and the forth is:

> Severity	Code	Description	Project	File	Line	Suppression State
> Error	CS1922	Cannot initialize type 'Vector<int>' with a collection initializer because it does not implement 'System.Collections.IEnumerable'	WindowsFormsApp6	C:\Users\mhg\Documents\Visual Studio 2017\Projects\WindowsFormsApp6\WindowsFormsApp6\Form1.cs	38	Active

**Comment by CyrusNajmabadi** [comment]
The parser is parsing that as:  ```var v = new Vector<int>() { 1, 2, 3, 4 } // and then );

The binder is then giving you the error based on the presumption you were typing a collection initializer incorrectly.

**Comment by ghost** [comment]
C# in general refuses this syntax:
```
void Foo (int[] A)
{

}

```
When I use:
`   Foo({ 1, 2, 3, 4 });`
It gives me 11 stntax erros, about missing sympols!
This is not about vectors in particular, but I was working with a vector sample.

**Comment by CyrusNajmabadi** [comment]
Yes.  We know that that is not allowed.  The question is:  If it were to be allowed, what would it mean?

I've given reasons for why i think it should *not* mean "Foo(new [] { 1, 2, 3, 4 })".

**Comment by CyrusNajmabadi** [comment]
> It gives me 11 stntax erros, about missing sympols

That is an issue with the compiler.   You are typing something it has no concept of, and it has to try to go figure out what you mean.  It sees an open curly, and thinks that maybe you meant a block, (or a collection initializer, or whatever).  You didn't, so things still keep going awry.  It's teh same as if i type any random piece of code i want to work into any compiler.  It will likely end up with many syntax errors.  It's not really relevant or important in this discussion.

**Comment by ghost** [comment]
@CyrusNajmabadi
I agree with you. The array type can be infered from the param. declaration. It is already working in VB.NET for many years. In the case of Vectors, the array type is also mentioned in the vector type parameter!.. Too much info!

**Comment by CyrusNajmabadi** [comment]
Ok.  So this seems like an extension to 'target typing'.  Where 'array literal' syntax should be allowed wherever a target type can be inferred?  If so, can you update your proposal accordingly and link to the other topics on this matter?

**Comment by ghost** [comment]
@CyrusNajmabadi 
> It's not really relevant or important in this discussion.

I think it is, becuase it explains where the "collection initializer " error came from. The compiler thinks that there are some missing symples, and the {1, 2, 3, 4} term is a collection initializer after them!

**Comment by ghost** [comment]
@CyrusNajmabadi 
> Ok. So this seems like an extension to 'target typing'. Where 'array literal' syntax should be allowed wherever a target type can be inferred? If so, can you update your proposal accordingly and link to the other topics on this matter?

Sorry, I'm not an expert. Could you do this in a reply, and I will copy it into my issue?

**Comment by CyrusNajmabadi** [comment]
> I think it is, becuase it explains where the "collection initializer " error came from. The compiler thinks that there are some missing symples, and the {1, 2, 3, 4} term is a collection initializer after them!

That is a concern abotu the C# compiler.  This is the repo for the C# language.  The error messages the compiler gives when it runs into things it doesn't understand are not relevant here.  This repo is for discussing what should/shouldn't be allowed in the language :)

**Comment by CyrusNajmabadi** [comment]
> Sorry, I'm not an expert. Could you do this in a reply, and I will copy it into my issue?

No :)   There are some decent docs on the site you could look at.  As well as previous examples of things that people have proposed.  It would also be good as it would give you a chance to read the relevant parts of the langauge spec so you could figure out what would be involved here. :)

**Comment by CyrusNajmabadi** [comment]
For example, this would help you understand how array initializers work, including the difference between an *target typed* array initializer, vs an *implicitly* typed array initializer. This understanding would help when creating these proposals.  It will also help indicate a level of seriousness behind understanding what is being asked for, and how it would fit into the language.

**Comment by ghost** [comment]
Thanks. This can happen by time, becuase I'm busy now.
But, I'll will be surprised if no one ever asked for this feature? May be dozens did this at other Microsoft feedback sites like "user voice".
I felt this urgent specially with vectors, because I must supply the type pram. because without it the compiler thinks it is the static Vector class!
This code seems too long (JUST A SAMPLE):
`var Result = Vector.Add(new Vector<short>(new short[] { 1, 2, 3, 4, 5, 6, 7, 8 }), new Vector<short>(new short[] { 1, 2, 3, 4, 5, 6, 7, 8 }));  
`
I dream it can be:
`var Result = Vector.Add<short>(<1, 2, 3, 4, 5, 6, 7, 8 >, <1, 2, 3, 4, 5, 6, 7, 8 >);  `

Or at least:
`var Result = Vector.Add(new Vector<short>({ 1, 2, 3, 4, 5, 6, 7, 8 }), new Vector<short>({ 1, 2, 3, 4, 5, 6, 7, 8 }));  `

**Comment by CyrusNajmabadi** [comment]
> But, I'll will be surprised if no one ever asked for this feature?

People ask for tons of language features :)  Everyone wants the language to be better for whatever specific coding task tehy're up to that day :)

**Comment by ghost** [comment]
And with some compiler work, it can be infere the Vector structuer type parameter from the method type parameter, without confusion with the vector class.. I think this syntax can be possible:
`var Result = Vector.Add<short>(new Vector({ 1, 2, 3, 4, 5, 6, 7, 8 }), new Vector({ 1, 2, 3, 4, 5, 6, 7, 8 }));`

**Comment by ghost** [comment]
By the way, I think things would be easier if the Vector class was named "Vectors" instead, so compilers wouldn't confuse it with the vector< T > struct!
Of course this is not the place for this complaining, but I don't expect they will accept renaming the class now!
So, C# and VB.NET compilers may carry out the mission here!

**Comment by ghost** [comment]
The same issue here:
`var Names = { "Ahmad", "Ali", "Abdo" };`
This will not compile, becuase C# can't infere that Names is a string array!
This is accepted in VB.NET:
`Dim Names = {"Ahmad", "Ali", "Abdo"}`

**Comment by ghost** [comment]
@MohammadHamdyGhanem hello!

> Why don't C# do the trick and convert
> { 1, 2, 3, 4 }
> to
> new int[]{ 1, 2, 3, 4 }
> based on the type of the parameter?
> Visual Basic Does this, and this code will compile:
> Dim v = New Vector(Of Integer)({1, 2, 3, 4})
> 
> The same issue here:
> var Names = { "Ahmad", "Ali", "Abdo" };
> This will not compile, becuase C# can't infere that Names is a string array!
> This is accepted in VB.NET:
> Dim Names = {"Ahmad", "Ali", "Abdo"}

If you want to use type infer for Array creation you may use following extensions
```c#
var array = New.Array(1, 2, 3, 4, 5);
var list = New.List(1, 2, 3, 4, 5);

var names = New.Array( "Ahmad", "Ali", "Abdo");

var keyToValue = New.Dictionary(
	"key0".To("value0"),
	"key1".To("value1"),
	"value2".By("key2")
);

var x = New.Object(type, constructorArgs);
var y = New.Object<T>(constructorArgs);
```
[See an online sample for your code](https://dotnetfiddle.net/KIPlF2)

Implementation
```c#
public static class New
{	
	public static T Object<T>() where T : new() => new T();

	public static T Object<T>(params object[] constructorArgs) =>
		(T) Activator.CreateInstance(typeof(T), constructorArgs);

	public static object Object(Type type, params object[] constructorArgs) =>
		Activator.CreateInstance(type, constructorArgs);
		
	public static T[] Array<T>(params T[] items) => items;
		
	public static List<T> List<T>(params T[] items) => new List<T>(items);

	public static Dictionary<TK, TV> Dictionary<TK, TV>(params KeyValuePair<TK, TV>[] items) =>
		new Dictionary<TK, TV>().Merge(items);
}

public static KeyValuePair<TK, TV> To<TK, TV>(this TK key, TV value) =>
	new KeyValuePair<TK, TV>(key, value);

public static KeyValuePair<TK, TV> By<TK, TV>(this TV value, TK key) =>
	new KeyValuePair<TK, TV>(key, value);
```
Some small benefits:
• allows to avoid additional `using` directives at some cases via type infer
• a convenient typing workflow
• a more abstract way with good readability

Also it is possible to create Merge extension for collections
```c#
var list = New.List(1, 2, 3, 4, 5);
Print(list); /* 1, 2, 3, 4, 5 */

list.Merge(6, 7, 8);
Print(list); /* 1, 2, 3, 4, 5, 6, 7, 8 */
```
Implementation
```c#
public static TCollection Merge<TCollection, T>(this TCollection collection, IEnumerable<T> items)
	where TCollection : ICollection<T>
{
	items.ForEach(collection.Add); // foreach (var item in items) collection.Add(item);
	return collection;
}
		
public static TCollection Merge<TCollection, T>(this TCollection collection, params T[] items)
	where TCollection : ICollection<T>
{
	items.ForEach(collection.Add); // foreach (var item in items) collection.Add(item);
	return collection;
}
```

Much more interesting syntax sugar for C# you may found by links
https://github.com/Makeloft/Ace/blob/master/Ace.Base/Sugar.cs
https://github.com/Makeloft/Ace/issues/2

**Comment by DarthVella** [comment]
Starting to wonder if Makeloft gets a commission each time someone uses their code.

Back on topic, even though it's probably not a good fit for C# I sometimes miss the simplicity of Javascript/Typescript's method of creating a filled array - just enclosing some items in square brackets.
```typescript
var array = [2.5, myDouble, 7.5];
```

**Reply by znakeeye** [reply]
> Back on topic, even though it's probably not a good fit for C# I sometimes miss the simplicity of Javascript/Typescript's method of creating a filled array - just enclosing some items in square brackets.
> 
> ```ts
> var array = [2.5, myDouble, 7.5];
> ```

I don't know how many times I have encountered a method that expects an array/enumerable and I have a single element. Certainly would be nice to allow for Javascript/Typescript syntax: **`Foo([myElement])`**
```c#
// C# 10.0 - too many tokens for my eyes...
Foo(new[] { myElement });

// C# 17.0 - meet the Typescript syntax! Also allows for compile-time optimizations, avoiding the allocation of an array.
Foo([myElement]);

void Foo(int[] items) {}
```

**Reply by HaloFour** [reply]
#5354

**Comment by Neme12** [comment]
@MohammadHamdyGhanem 

> The same issue here:
var Names = { "Ahmad", "Ali", "Abdo" };

Under the 'target typed array initializer' proposal, that wouldn't work either. As @CyrusNajmabadi said, please think a little about what your proposal actually is and how the feature should work. Saying "I wish this worked, and something else too, and the compiler should magically figure out what I want" is nice, but won't get us anywhere. Especially if you don't explain *how* it would do that.

> This is accepted in VB.NET:
Dim Names = {"Ahmad", "Ali", "Abdo"}

You mentioned VB quite a few times. If you prefer it to C#, may I ask why you don't continue using it? We are definitely not going to put all VB features in C#. Sometimes, a feature isn't missing just because "it's missing" or because it's an omission. Often it's by design.

@DarthVella 

> not a good fit for C#

Correct.

**Comment by Neme12** [comment]
As for having a target-typed array initializer, personally I wouldn't be opposed to it. In fact at one point I used to think this feature already existed. The array initializer syntax sort of gives you that impression:
```c#
int[] array = { 1, 2, 3 };
```
I think it would be a natural generalization of this. But I can also live without it.

**Comment by ghost** [comment]
> You mentioned VB quite a few times. If you prefer it to C#, may I ask why you don't continue using it? We are definitely not going to put all VB features in C#. Sometimes, a feature isn't missing just because "it's missing" or because it's an omission. Often it's by design.

I use both. Microsoft forces Vb programers to use C# in. Net core apps.
So many times I get syntax errors in C#  becuase I am using the  easier array initialization syntax that works in VB.  Besides I mention VB to say that the syntax aleady exists somewhere and is not a strage concept.
I don't see the need for more explanation of how the syntax is going to work. If it doesn,t cause any problems, which no one mentioned so far,  then C# team has the ability to make this syntax work.

**Comment by CyrusNajmabadi** [comment]
>  I  don't see the need for more explanation of how the syntax is going to  work. If it doesn,t cause any problems, which no one mentioned so far,   then C# team has the ability to make this syntax work.

What i would appreciate is a more fully fleshed out proposal as to what the feature actually is, as well as the pros/cons around it.  Simply stating that you want something without a good description of what that want is is not that helpful.

**Comment by Austin-bryan** [comment]
@MohammadHamdyGhanem 
> It gives me 11 stntax erros, about missing sympols!

It gives me 11 syntax errors, about missing symbols!*

At least try to type...
